composer/utils/fx_utils.py

```diff
@@ -123,16 +149,106 @@ def detect_residual_pattern(gm: GraphModule):
     raise NotImplementedError('detect_residual_pattern is currently not implemented.')
 
 
-def replace_residual_with_stochastic(gm: GraphModule):
-    """Replaces residual pattern with their stoachstic equivalent.
+def _get_ancestors(node: Node) -> List[Node]:
+    ancestorNodes = []
+    while node.op != 'placeholder':
+        ancestorNodes.append(node)
+        node = node.all_input_nodes[0]
+    return ancestorNodes
+
+
+def _get_residual_path(nodeLHS: Node, nodeRHS: Node) -> Tuple[List[Node], List[Node]]:
+    """Walk backwards from nodeLHS and nodeRSH to the root and construct lists of their parents.
 
     Arguments:
-        gm (GraphModule): The source FX-traced graph.
+        nodeLHS (Node): left-hand side node for a binary operator
+        nodeRHS (Node): right-hand side node for a binary operator
 
     Returns:
-        GraphModule: Modified GraphModule.
+        (lhsAncestors, rhsAncesotors): Two lists of nodes containing ancestors for ``nodeLHS`` and ``nodeRHS`` with
+            their common ancestors removed.
     """
-    raise NotImplementedError('replace_residual_with_stochastic is currently not implemented.')
+    lhsAncestors = _get_ancestors(nodeLHS)
+    rhsAncesotors = _get_ancestors(nodeRHS)
+
+    # Iterate from back and eliminate common nodes
+    while lhsAncestors and rhsAncesotors and lhsAncestors[-1] == rhsAncesotors[-1]:  # type: ignore
+        del lhsAncestors[-1]
+        del rhsAncesotors[-1]
+    lhsAncestors.reverse()
+    rhsAncesotors.reverse()
+    return lhsAncestors, rhsAncesotors
+
+
+def _tag_nodes(nodes: List[Node], tag: str):
+    """Tag nodes for the splitter."""
+    for node in nodes:
+        node.tag = tag  # type: ignore[attr-defined]
+
+
+def apply_stochastic_residual(gm: GraphModule, drop_rate: float = 0.2) -> Tuple[GraphModule, int]:
+    """Detect and replace residual pattern with their stochastic equivalent.
+
+    Arguments:
+        gm (GraphModule): The source FX-traced graph. It can be the whole model symbolically traced.
+
+    Returns:
+        GraphModule: Modified GraphModule that has stochastic residual connections.
+    """
+    assert isinstance(gm, GraphModule), 'Input to apply_stochastic_residual should be an instance of GraphModule'
```

Comment:
No need for this assert -- it is implied by the type annotation

Reply:
That just a static check and this API is used in non-type checked setting it will throw an unrelated error.

Reply:
```
>>> from composer.utils.fx_utils import apply_stochastic_residual
>>> x = 10
>>> apply_stochastic_residual(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workdisk/daya/composer/composer/utils/fx_utils.py", line 209, in apply_stochastic_residual
    assert isinstance(gm, GraphModule), 'Input to apply_stochastic_residual should be an instance of GraphModule'
AssertionError: Input to apply_stochastic_residual should be an instance of GraphModule
```

vs 

```
>>> from composer.utils.fx_utils import apply_stochastic_residual
>>> x = 10
>>> apply_stochastic_residual(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workdisk/daya/composer/composer/utils/fx_utils.py", line 216, in apply_stochastic_residual
    for node in gm.graph.nodes:
AttributeError: 'int' object has no attribute 'graph'
>>>
```